base/essentials.jl

```diff
@@ -813,8 +813,7 @@ const missing = Missing()
 
 Indicate whether `x` is [`missing`](@ref).
 """
-ismissing(::Any) = false
-ismissing(::Missing) = true
+ismissing(x) = x === missing
```

Comment:
Hmm, I'm not sure 

> This seems somewhat harder to infer and cache, though we want it to be equivalent.

Could you please elaborate this ?

I included this change (as I did for `isnothing`) considering we can benefit from this with some code like below
```julia
code_typed((Vector{Union{Missing,Int}},); optimize = false) do v
    s = 0.
    for a in v
        ismissing(a) && continue
        s + sin(a) # type stable, `a::Int`
    end
    s
end |> first
```

Reply:
Ah, right, because we don't infer it due to the early union-splitting. I think we should? (it's quite similar to the `ispositive` case also in this framework)

Reply:
yes, that's one of the current limitations. For now `Conditional`s never appear in code like below:
```julia
checker(@nospecialize(_)) = false
checker(::SomeType1) = true
checker(::SomeType2) = true
```

Reply:
What happens if a package defines `Base.ismissing(::MyCustomMissing) = true`? This is something that could be useful at some point. Would that break optimizations?

Reply:
The type constraint will be there as far as the number of `ismissing` methods is [≤3](https://github.com/JuliaLang/julia/blob/5c6e21edbfd8f0c7d16ea01c91d1c75c30d4eaa1/base/compiler/types.jl#L119)
```julia
julia> struct MyCustomMissing end

julia> Base.ismissing(::MyCustomMissing) = true

julia> code_typed((Any,)) do a
           ismissing(a) && return a
           0
       end
1-element Vector{Any}:
 CodeInfo(
1 ─ %1 = (isa)(a, MyCustomMissing)::Bool
└──      goto #3 if not %1
2 ─      goto #4
3 ─ %4 = Main.ismissing(a)::Bool
└──      goto #4
4 ┄ %6 = φ (#2 => true, #3 => %4)::Bool
└──      goto #6 if not %6
5 ─ %8 = π (a, Union{Missing, MyCustomMissing})
└──      return %8
6 ─      return 0
) => Union{Missing, MyCustomMissing, Int64}
```